Error while deserializing sketches sent from ASAP summary ingest to ASAP query engine

### Steps to Reproduce Error

Working Branch: `verify-sketchlib-integration`
All Python scripts are to be run from the `~/ASAPQuery/asap-tools/execution-utilities/benchmark` directory.

#### Install Elastic
`curl -fsSL https://elastic.co/start-local | sh`

This should output an API key, replace `your-api-key` with that value in all commands below.

#### Download data
`python download_dataset.py --dataset h2o --output-dir ./data`

#### Export to database
`python export_to_database.py --dataset h2o --file-path ./data/G1_1e7_1e2_0_0.csv --es-host localhost --es-port 9200 --es-index h2o_groupby --es-api-key your-api-key`

#### Prepare data for Arroyo
`python prepare_data.py --dataset h2o --input ./data/G1_1e7_1e2_0_0.csv --output ./data/h2o_arroyo.parquet`

#### Generate query files
`python generate_queries.py --table-name h2o_groupby --ts-column timestamp --value-column v3 --group-by-columns id1,id2 --window-size 10 --num-queries 50 --auto-detect-timestamps --data-file ./data/h2o_arroyo.parquet --data-file-format json --output-prefix ./queries/h2o --stride-seconds 10`

#### Kafka and Arroyo
Start Kafka:
```
cd kafka
./bin/kafka-server-start.sh ./config/kraft/server.properties
./bin/kafka-topics.sh --bootstrap-server localhost:9092 --create --topic sketch_topic --partitions 1 --replication-factor 1 --config max.message.bytes=20971520 --config [retention.ms](http://retention.ms/)=-1
```

Start Arroyo:
```
./target/release/arroyo --config ~/ASAPQuery/asap-summary-ingest/config.yaml cluster
```

#### Export data to Arroyo
`python export_to_arroyo.py  --streaming-config configs/h2o_streaming.yaml  --source-type file     --input-file ./data/h2o_arroyo.parquet  --file-format parquet  --ts-format unix_millis --pipeline-name h2o_pipeline  --arroyosketch-dir ~/ASAPQuery/asap-summary-ingest`

#### Start Query Engine
1. Without legacy flag:
`./target/release/query_engine_rust --kafka-topic sketch_topic --input-format json --prometheus-scrape-interval 1 --config ~/ASAPQuery/asap-tools/execution-utilities/benchmark/h2o_inference.yaml --streaming-config ~/ASAPQuery/asap-tools/execution-utilities/benchmark/h2o_streaming.yaml --prometheus-server http://localhost:9200 --http-port 8088 --delete-existing-db --log-level info --output-dir ./output --streaming-engine arroyo --query-language ElasticSQL --lock-strategy per-key`

```
2026-04-07T11:24:36.363289Z ERROR query_engine_rust::drivers::ingest::kafka: asap-query-engine/src/drivers/ingest/kafka.rs:316: Error deserializing Arroyo PrecomputedOutput from JSON with precompute: Failed to deserialize DatasketchesKLLAccumulator: invalid type: floating point 59.387454, expected a sequence
2026-04-07T11:24:36.363321Z ERROR query_engine_rust::drivers::ingest::kafka: asap-query-engine/src/drivers/ingest/kafka.rs:146: Error processing message: Failed to deserialize DatasketchesKLLAccumulator: invalid type: floating point 59.387454, expected a sequence
```

2. With legacy flag (`--sketch-kll-impl legacy`)
```
2026-04-08T13:22:44.499871Z ERROR query_engine_rust::drivers::ingest::kafka: asap-query-engine/src/drivers/ingest/kafka.rs:146: Error processing message: Failed to deserialize DatasketchesKLLAccumulator: Failed to deserialize KLL sketch: Error: Possible corruption: M must be 8: 187
```

--- a/asap-summary-ingest/run_arroyosketch.py
+++ b/asap-summary-ingest/run_arroyosketch.py
@@ -952,6 +952,10 @@ def main(args):
             parameters["impl_mode"] = getattr(
                 args, "sketch_cmwh_impl", "legacy"
             ).capitalize()
+        elif agg_function in ("datasketcheskll_", "hydrakll_"):
+            parameters["impl_mode"] = getattr(
+                args, "sketch_kll_impl", "sketchlib"
+            ).capitalize()
 
         sql_queries.append(sql_query)
         # if not is_labels_accumulator:

--- a/asap-summary-ingest/templates/udfs/datasketcheskll_.rs.j2
+++ b/asap-summary-ingest/templates/udfs/datasketcheskll_.rs.j2
@@ -1,5 +1,6 @@
 /*
 [dependencies]
+dsrs = { git = "https://github.com/ProjectASAP/datasketches-rs", rev = "d748ec75c80fff21f7b24897244dd1c895df2e9a" }
 asap_sketchlib = { git = "https://github.com/ProjectASAP/asap_sketchlib" }
 arroyo-udf-plugin = "0.1"
 rmp-serde = "1.1"
@@ -7,10 +8,23 @@ serde = { version = "1.0", features = ["derive"] }
 */
 
 use arroyo_udf_plugin::udf;
+use dsrs::KllDoubleSketch;
 use serde::{Deserialize, Serialize};
 use asap_sketchlib::KLL;
 
 const DEFAULT_K: u16 = {{ k }};
+
+enum ImplMode {
+    Legacy,
+    Sketchlib,
+}
+
+{% set _impl_mode = impl_mode | default("Sketchlib") %}
+const IMPL_MODE: ImplMode = ImplMode::{% if _impl_mode == "Legacy" or _impl_mode == "Sketchlib" %}{{ _impl_mode }}{% else %}Sketchlib{% endif %};
+
+fn use_sketchlib_for_kll() -> bool {
+    matches!(IMPL_MODE, ImplMode::Sketchlib)
+}
 
 #[derive(Serialize, Deserialize)]
 struct KllSketchData {
@@ -20,11 +34,19 @@ struct KllSketchData {
 
 #[udf]
 fn datasketcheskll_(values: Vec<f64>) -> Option<Vec<u8>> {
-    let mut sketch = KLL::init_kll(DEFAULT_K as i32);
-    for &value in &values {
-        sketch.update(&value);
-    }
-    let sketch_bytes = sketch.serialize_to_bytes().ok()?;
+    let sketch_bytes = if use_sketchlib_for_kll() {
+        let mut sketch = KLL::init_kll(DEFAULT_K as i32);
+        for &value in &values {
+            sketch.update(&value);
+        }
+        sketch.serialize_to_bytes().ok()?
+    } else {
+        let mut sketch = KllDoubleSketch::with_k(DEFAULT_K);
+        for &value in &values {
+            sketch.update(value);
+        }
+        sketch.serialize().as_ref().to_vec()
+    };
     let serialized = KllSketchData {
         k: DEFAULT_K,
         sketch_bytes,

--- a/asap-summary-ingest/templates/udfs/hydrakll_.rs.j2
+++ b/asap-summary-ingest/templates/udfs/hydrakll_.rs.j2
@@ -1,5 +1,6 @@
 /*
 [dependencies]
+dsrs = { git = "https://github.com/ProjectASAP/datasketches-rs", rev = "d748ec75c80fff21f7b24897244dd1c895df2e9a" }
 asap_sketchlib = { git = "https://github.com/ProjectASAP/asap_sketchlib" }
 arroyo-udf-plugin = "0.1"
 rmp-serde = "1.1"
@@ -8,10 +9,23 @@ xxhash-rust = { version = "0.8", features = ["xxh32"] }
 */
 
 use arroyo_udf_plugin::udf;
+use dsrs::KllDoubleSketch;
 use rmp_serde::Serializer;
 use serde::{Deserialize, Serialize};
 use asap_sketchlib::KLL;
 use xxhash_rust::xxh32::xxh32;
+
+enum ImplMode {
+    Legacy,
+    Sketchlib,
+}
+
+{% set _impl_mode = impl_mode | default("Sketchlib") %}
+const IMPL_MODE: ImplMode = ImplMode::{% if _impl_mode == "Legacy" or _impl_mode == "Sketchlib" %}{{ _impl_mode }}{% else %}Sketchlib{% endif %};
+
+fn use_sketchlib_for_kll() -> bool {
+    matches!(IMPL_MODE, ImplMode::Sketchlib)
+}
 
 const ROW_NUM: usize = {{ row_num }};
 const COL_NUM: usize = {{ col_num }};
@@ -32,40 +46,76 @@ struct HydraKllSketchData {
 
 #[udf]
 fn hydrakll_(keys: Vec<&str>, values: Vec<f64>) -> Option<Vec<u8>> {
-    let mut sketches: Vec<Vec<KLL>> = (0..ROW_NUM)
-        .map(|_| {
-            (0..COL_NUM)
-                .map(|_| KLL::init_kll(DEFAULT_K as i32))
-                .collect()
-        })
-        .collect();
+    let sketch_data: Vec<Vec<KllSketchData>> = if use_sketchlib_for_kll() {
+        let mut sketches: Vec<Vec<KLL>> = (0..ROW_NUM)
+            .map(|_| {
+                (0..COL_NUM)
+                    .map(|_| KLL::init_kll(DEFAULT_K as i32))
+                    .collect()
+            })
+            .collect();
 
-    for (i, &key) in keys.iter().enumerate() {
-        if i >= values.len() {
-            break;
+        for (i, &key) in keys.iter().enumerate() {
+            if i >= values.len() {
+                break;
+            }
+            let key_bytes = key.as_bytes();
+            for row in 0..ROW_NUM {
+                let hash_value = xxh32(key_bytes, row as u32);
+                let col_index = (hash_value as usize) % COL_NUM;
+                sketches[row][col_index].update(&values[i]);
+            }
         }
-        let key_bytes = key.as_bytes();
-        for row in 0..ROW_NUM {
-            let hash_value = xxh32(key_bytes, row as u32);
-            let col_index = (hash_value as usize) % COL_NUM;
-            sketches[row][col_index].update(&values[i]);
-        }
-    }
 
-    let sketch_data: Vec<Vec<KllSketchData>> = sketches
-        .iter()
-        .map(|row| {
-            row.iter()
-                .map(|sketch| {
-                    let sketch_bytes = sketch.serialize_to_bytes().ok()?;
-                    Some(KllSketchData {
-                        k: DEFAULT_K,
-                        sketch_bytes,
+        sketches
+            .iter()
+            .map(|row| {
+                row.iter()
+                    .map(|sketch| {
+                        let sketch_bytes = sketch.serialize_to_bytes().ok()?;
+                        Some(KllSketchData {
+                            k: DEFAULT_K,
+                            sketch_bytes,
+                        })
                     })
-                })
-                .collect::<Option<Vec<_>>>()
-        })
-        .collect::<Option<Vec<_>>>()?;
+                    .collect::<Option<Vec<_>>>()
+            })
+            .collect::<Option<Vec<_>>>()?
+    } else {
+        let mut sketches: Vec<Vec<KllDoubleSketch>> = (0..ROW_NUM)
+            .map(|_| {
+                (0..COL_NUM)
+                    .map(|_| KllDoubleSketch::with_k(DEFAULT_K))
+                    .collect()
+            })
+            .collect();
+
+        for (i, &key) in keys.iter().enumerate() {
+            if i >= values.len() {
+                break;
+            }
+            let key_bytes = key.as_bytes();
+            for row in 0..ROW_NUM {
+                let hash_value = xxh32(key_bytes, row as u32);
+                let col_index = (hash_value as usize) % COL_NUM;
+                sketches[row][col_index].update(values[i]);
+            }
+        }
+
+        sketches
+            .iter()
+            .map(|row| {
+                row.iter()
+                    .map(|sketch| {
+                        Some(KllSketchData {
+                            k: DEFAULT_K,
+                            sketch_bytes: sketch.serialize().as_ref().to_vec(),
+                        })
+                    })
+                    .collect::<Option<Vec<_>>>()
+            })
+            .collect::<Option<Vec<_>>>()?
+    };
 
     let hydra_data = HydraKllSketchData {
         row_num: ROW_NUM,

--- a/asap-summary-ingest/tests/test_integration.py
+++ b/asap-summary-ingest/tests/test_integration.py
@@ -430,5 +430,48 @@ metrics:
         agg_configs[0].validate(metric_config, query_language="promql")
 
 
+class TestKllUdfTemplates:
+    """Tests for KLL UDF template rendering."""
+
+    @pytest.fixture
+    def udf_template_dir(self):
+        """Load the UDF template directory."""
+        return os.path.join(
+            os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
+            "templates",
+            "udfs",
+        )
+
+    def test_template_variables_ignore_local_jinja_assignments(self):
+        """Test that local Jinja variables are not treated as required params."""
+        template_source = """
+{% set _impl_mode = impl_mode | default("Sketchlib") %}
+{{ _impl_mode }} {{ k }}
+"""
+        template_vars = jinja_utils.get_template_variables(template_source)
+
+        assert template_vars == {"impl_mode", "k"}
+
+    def test_datasketches_kll_template_renders_legacy_mode(self, udf_template_dir):
+        """Test rendering the single KLL UDF with the legacy backend."""
+        template = jinja_utils.load_template(udf_template_dir, "datasketcheskll_.rs.j2")
+        rendered = template.render(k=200, impl_mode="Legacy")
+
+        assert "const IMPL_MODE: ImplMode = ImplMode::Legacy;" in rendered
+        assert "KllDoubleSketch::with_k(DEFAULT_K)" in rendered
+        assert "asap_sketchlib" in rendered
+        assert "{{" not in rendered
+
+    def test_hydra_kll_template_renders_sketchlib_mode(self, udf_template_dir):
+        """Test rendering the Hydra KLL UDF with the sketchlib backend."""
+        template = jinja_utils.load_template(udf_template_dir, "hydrakll_.rs.j2")
+        rendered = template.render(row_num=3, col_num=128, k=200, impl_mode="Sketchlib")
+
+        assert "const IMPL_MODE: ImplMode = ImplMode::Sketchlib;" in rendered
+        assert "KLL::init_kll(DEFAULT_K as i32)" in rendered
+        assert "KllDoubleSketch::with_k(DEFAULT_K)" in rendered
+        assert "{{" not in rendered
+
+
 if __name__ == "__main__":
     pytest.main([__file__, "-v"])

--- a/asap-summary-ingest/utils/jinja_utils.py
+++ b/asap-summary-ingest/utils/jinja_utils.py
@@ -1,4 +1,4 @@
-from jinja2 import Environment, FileSystemLoader, nodes
+from jinja2 import Environment, FileSystemLoader, meta
 
 
 def load_template(template_dir, template_name):
@@ -23,5 +23,4 @@ def get_template_variables(template_source, environment=None):
         environment = Environment()
 
     ast = environment.parse(template_source)
-    template_vars = ast.find_all(nodes.Name)
-    return {var.name for var in template_vars if var.ctx == "load"}
+    return meta.find_undeclared_variables(ast)

